cpp/include/kvikio/shim/libcurl.hpp

```diff
@@ -0,0 +1,260 @@
+    if ((ver->features & CURL_VERSION_THREADSAFE) == 0) {
+      throw std::runtime_error("cannot initialize libcurl - built with thread safety disabled");
+    }
+  }
+  ~LibCurl() noexcept
+  {
+    _free_curl_handles.clear();
+    curl_global_cleanup();
+  }
+
+ public:
+  static LibCurl& instance()
+  {
+    static LibCurl _instance;
+    return _instance;
+  }
+
+  /**
+   * @brief Returns a free curl handle if available.
+   */
+  UniqueHandlePtr get_free_handle()
+  {
+    UniqueHandlePtr ret;
+    std::lock_guard const lock(_mutex);
+    if (!_free_curl_handles.empty()) {
+      ret = std::move(_free_curl_handles.back());
+      _free_curl_handles.pop_back();
+    }
+    return ret;
+  }
+
+  /**
+   * @brief Returns a curl handle, create a new handle if none is available.
+   */
+  UniqueHandlePtr get_handle()
+  {
+    // Check if we have a free handle available.
+    UniqueHandlePtr ret = get_free_handle();
+    if (ret) {
+      curl_easy_reset(ret.get());
+    } else {
+      // If not, we create a new handle.
+      CURL* raw_handle = curl_easy_init();
+      if (raw_handle == nullptr) {
+        throw std::runtime_error("libcurl: call to curl_easy_init() failed");
+      }
+      ret = UniqueHandlePtr(raw_handle, curl_easy_cleanup);
+    }
+    return ret;
+  }
+
+  /**
+   * @brief Retain a curl handle for later use.
+   */
+  void retain_handle(UniqueHandlePtr handle)
+  {
+    std::lock_guard const lock(_mutex);
+    _free_curl_handles.push_back(std::move(handle));
+  }
+};
+
+/**
+ * @brief Representation of a curl easy handle pointer and its operations.
+ *
+ * An instance is given a `LibCurl::UniqueHandlePtr` on creation, which is
+ * later retailed on destruction.
+ */
+class CurlHandle {
+ private:
+  char _errbuf[CURL_ERROR_SIZE];
+  LibCurl::UniqueHandlePtr _handle;
+  std::string _source_file;
+  std::string _source_line;
+
+ public:
+  /**
+   * @brief Construct a new curl handle.
+   *
+   * Typically, do not use this directly instead use the `create_curl_handle()` macro.
+   *
+   * @param handle An unused curl easy handle pointer, which is retailed on destruction.
+   * @param source_file Path of source file of the caller (for error messages).
+   * @param source_line Line of source file of the caller (for error messages).
+   */
+  CurlHandle(LibCurl::UniqueHandlePtr handle, std::string source_file, std::string source_line)
+    : _handle{std::move(handle)},
+      _source_file(std::move(source_file)),
+      _source_line(std::move(source_line))
+  {
+    // Need CURLOPT_NOSIGNAL to support threading, see
+    // <https://curl.se/libcurl/c/CURLOPT_NOSIGNAL.html>
+    setopt(CURLOPT_NOSIGNAL, 1L);
+
+    // We always set CURLOPT_ERRORBUFFER to get better error messages.
+    setopt(CURLOPT_ERRORBUFFER, _errbuf);
+
+    // Make curl_easy_perform() fail when receiving HTTP code errors.
+    setopt(CURLOPT_FAILONERROR, 1L);
+  }
+  ~CurlHandle() noexcept { LibCurl::instance().retain_handle(std::move(_handle)); }
```

Comment:
*question*: One thing I wonder is whether this is ahead-of-time optimisation, or if this mechanism of keeping hold of curl handles is a recommended pattern?

Reply:
I haven't benchmark it but the docs says:
> The easy handle is used to hold and control a single network transfer. It is encouraged to reuse easy handles for repeated transfers.

Reply:
https://curl.se/libcurl/c/curl_easy_init.html

Reply:
Thanks.